src/Lykke.Service.TradeVolumes.AzureRepositories/TradeVolumesRepository.cs

```diff
@@ -42,81 +60,199 @@ public TradeVolumesRepository(IReloadingManager<string> connectionStringManager,
-                baseVolume,
-                dateTime);
-            var quotingStorage = GetStorage(quotingAssetId, dateTime);
-            await quotingStorage.InsertOrMergeAsync(quotingEntity);
         }
 
         public async Task<double> GetPeriodClientVolumeAsync(
             string baseAssetId,
             string quotingAssetId,
             string clientId,
             DateTime from,
-            DateTime to)
+            DateTime to,
+            string excludeClientId)
         {
-            double result = 0;
-            for (DateTime start = from.Date; start < to.Date; start = start.AddDays(1))
+            var tradeVolumes = new List<double>();
+            var months = new List<DateTime>();
+            var existingTables = new HashSet<string>();
+            for (DateTime start = from; start < to; start = start.AddMonths(1))
+            {
+                months.Add(start);
+                await AddExistingTableNamesAsync(baseAssetId, start, existingTables);
+            }
+            var dates = new List<DateTime>();
+            for (DateTime start = from; start < to; start = start.AddHours(1))
             {
-                double tradeVolume =  await GetTradeVolumeAsync(
-                    start,
+                var tableName = GetTableName(baseAssetId, start);
+                if (existingTables.Contains(tableName))
+                    dates.Add(start);
+            }
+            await Task.WhenAll(dates.Select(d =>
+                AddTradeVolumeAsync(
+                    d,
                     clientId,
                     baseAssetId,
-                    quotingAssetId);
-                result += tradeVolume;
-            }
-            return result;
+                    quotingAssetId,
+                    excludeClientId,
+                    tradeVolumes)));
+            return tradeVolumes.Sum();
         }
 
         public async Task<(double, double)> GetClientPairValuesAsync(
             DateTime date,
             string clientId,
             string baseAssetId,
-            string quotingAssetId)
+            string quotingAssetId,
+            string excludeClientId)
         {
             double baseTradeVolume = await GetTradeVolumeAsync(
                 date,
                 clientId,
                 baseAssetId,
-                quotingAssetId);
+                quotingAssetId,
+                excludeClientId);
 
             double quotingTradeVolume = await GetTradeVolumeAsync(
                 date,
                 clientId,
                 quotingAssetId,
-                baseAssetId);
+                baseAssetId,
+                excludeClientId);
 
             return (baseTradeVolume, quotingTradeVolume);
         }
 
+        private async Task AddTradeVolumeAsync(
+            DateTime date,
+            string clientId,
+            string baseAssetId,
+            string quotingAssetId,
+            string excludeClientId,
+            List<double> tradeVolumes)
```

Comment:
A ConcurrentBag class can be used here

Reply:
We don't need concurrent read operations here.

Reply:
It also works with a concurrent write with less overhead than _lock.WaitAsync(). Plus the code will be clearer.